keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ImageResolver.kt

```diff
@@ -0,0 +1,82 @@
+package com.netflix.spinnaker.keel.ec2.resource
+
+import com.netflix.spinnaker.keel.api.NoImageFound
+import com.netflix.spinnaker.keel.api.NoImageFoundForRegion
+import com.netflix.spinnaker.keel.api.NoImageSatisfiesConstraints
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.UnsupportedStrategy
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.api.ec2.image.ArtifactImageProvider
+import com.netflix.spinnaker.keel.api.ec2.image.IdImageProvider
+import com.netflix.spinnaker.keel.api.ec2.image.ImageProvider
+import com.netflix.spinnaker.keel.api.uid
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.clouddriver.model.NamedImageComparator
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.PromotionRepository
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import org.slf4j.LoggerFactory
+
+class ImageResolver(
+  private val dynamicConfigService: DynamicConfigService,
+  private val cloudDriverService: CloudDriverService,
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val promotionRepository: PromotionRepository
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  // TODO: we may (probably will) want to make the parameter type more generic
+  suspend fun resolveImageId(resource: Resource<ClusterSpec>): String {
+    when (val imageProvider = resource.spec.launchConfiguration.imageProvider) {
+      is IdImageProvider -> {
+        return imageProvider.imageId
+      }
+      is ArtifactImageProvider -> {
+        val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(resource.uid)
+        val environment = deliveryConfigRepository.environmentFor(resource.uid)
+        val artifact = imageProvider.deliveryArtifact
+        val artifactName = if (deliveryConfig != null && environment != null) {
+          val version = promotionRepository.latestVersionApprovedIn(
+            deliveryConfig,
+            artifact,
+            environment.name
+          ) ?: throw NoImageSatisfiesConstraints(artifact.name, environment.name)
+          "${artifact.name}-$version"
+        } else {
+          artifact.name
+        }
+        val account = dynamicConfigService.getConfig("images.default-account", "test")
+        val namedImage = cloudDriverService
+          .namedImages(artifactName, account)
+          .sortedWith(NamedImageComparator)
+          .firstOrNull() ?: throw NoImageFound(artifactName)
+
+        log.info("Image found for {}: {}", artifactName, namedImage)
+
+        return namedImage.amis[resource.spec.location.region]?.first()
+          ?: throw NoImageFoundForRegion(artifactName, resource.spec.location.region) // todo eb: when are there multiple?
+      }
+//      is JenkinsImageProvider -> {
```

Comment:
do you want this to be commented out?

Reply:
Damn, sorry. Meant to write some tests for this and forgot.

Reply:
I put it back (still needs tests!)